qutebrowser/browser/webkit/network/httpproxy.py

```diff
@@ -0,0 +1,123 @@
+"""Asynchronous resolving of HTTP proxies."""
+
+from PyQt5.QtCore import QObject, QEventLoop, pyqtSlot
+from PyQt5.QtNetwork import QNetworkProxy, QNetworkRequest, \
+    QNetworkReply, QNetworkAccessManager
+
+from qutebrowser.browser.webkit.network.pac import PACResolver, EvalProxyError
+from qutebrowser.utils import log
+
+
+def proxy_from_url(typ, url):
+    """Create a QNetworkProxy from QUrl and a proxy type.
+
+    Args:
+        typ: QNetworkProxy::ProxyType.
+        url: URL of a proxy (possibly with credentials).
+
+    Return:
+        New QNetworkProxy.
+    """
+    proxy = QNetworkProxy(typ, url.host())
+    if url.port() != -1:
+        proxy.setPort(url.port())
+    if url.userName():
+        proxy.setUser(url.userName())
+    if url.password():
+        proxy.setPassword(url.password())
+    return proxy
+
+
+class HTTPProxyResolver(QObject):
+
+    """Asynchronous resolver of HTTP proxies."""
+
+    def __init__(self, url):
+        """Resolve a HTTP proxy from URL.
+
+        Args:
+            url: QUrl of a HTTP proxy.
+        """
+        QObject.__init__(self)
+
+        assert url.scheme() == "http" or url.scheme() == "https"
+        if not url.fileName().endswith(".pac"):
+            self._fallback(url)
+        else:
+            self._manager = QNetworkAccessManager()
+            self._manager.setProxy(QNetworkProxy(QNetworkProxy.NoProxy))
+            self._reply = self._manager.get(QNetworkRequest(url))
+            self._reply.metaDataChanged.connect(self._check_metadata)
+            self._reply.finished.connect(self._finish)
+            self._pac = None
+
+    @pyqtSlot()
+    def _finish(self):
+        self._check_metadata()
+        if not hasattr(self, "_proxy"):
+            if self._reply.error() != QNetworkReply.NoError:
+                self._fallback(self._reply.url())
+            else:
+                pacscript = bytes(self._reply.readAll()).decode()
+                try:
+                    self._pac = PACResolver(pacscript)
+                    log.network.debug("Successfully evaluated PAC file.")
+                    self._reply.deleteLater()
+                except EvalProxyError as e:
+                    error = "Error in PAC evaluation: {!s}.".format(e)
+                    log.network.exception(error)
+                    self.fallback()
+
+    def _fallback(self, url):
+        log.network.debug("Giving up on using proxy URL as PAC file.")
+        self._proxy = proxy_from_url(QNetworkProxy.HttpProxy, url)
+        self._pac = None
+        if hasattr(self, "_reply"):
+            self._reply.deleteLater()
+
+    @pyqtSlot()
+    def _check_metadata(self):
+        typ = self._reply.header(QNetworkRequest.ContentTypeHeader)
+        if not (typ is None or typ == "application/x-ns-proxy-autoconfig"):
+            self._fallback(self._reply.url())
+
+    def is_pac(self):
+        """Return True iff PAC is used."""
+        if hasattr(self, "_manager"):
+            loop = QEventLoop()
```

Comment:
So as you mentioned yourself, having a local `QEventLoop` is really difficult to reason about (and can lead to obscure bugs), so I'd like to avoid that if possible in any way.

How could we do this here? I guess we could not attempt to guess whether something is a PAC file or not, and handle not being able to parse a PAC file in some other way (ideally by not letting qutebrowser connect anymore at all).